File: problems/05_binary_search/0704_binary_search/solution.py
"""
Problem: 0704. Binary Search
Difficulty: Easy
URL: https://leetcode.com/problems/binary-search/

解法演進：
1) 迭代二分（閉區間）：l=0, r=n-1；mid 比較。
2) 遞迴二分：相同語意。

最佳解核心思路：
- 每輪以 mid 將搜尋區間減半。

複雜度：O(log n) / O(1)。
"""

class Solution:
    def search(self, nums: list[int], target: int) -> int:
        pass

    def search_brute(self, nums: list[int], target: int) -> int:
        # 線性搜尋對照
        pass


if __name__ == "__main__":
    solution = Solution()

    # Test 1
    # result = solution.solve(...)
    # assert result == expected

    print("All tests passed!")

File: problems/05_binary_search/0704_binary_search/solution.swift
// Problem: 0704. Binary Search
// Difficulty: Easy

import Foundation

class Solution {
    func search(_ nums: [Int], _ target: Int) -> Int {
        var l = 0, r = nums.count - 1
        while l <= r {
            let mid = l + (r - l) / 2
            if nums[mid] == target { return mid }
            if nums[mid] < target { l = mid + 1 } else { r = mid - 1 }
        }
        return -1
    }
}

func checkInt(_ a: Int, _ e: Int, _ msg: String) {
    if a != e { fatalError("FAIL \(msg): \(a) vs \(e)") }
}

let sol = Solution()
checkInt(sol.search([-1, 0, 3, 5, 9, 12], 9), 4, "ex1")
checkInt(sol.search([-1, 0, 3, 5, 9, 12], 2), -1, "ex2")
checkInt(sol.search([5], 5), 0, "single hit")
checkInt(sol.search([5], 3), -1, "single miss")
checkInt(sol.search([1, 2, 3, 4, 5], 1), 0, "first")
checkInt(sol.search([1, 2, 3, 4, 5], 5), 4, "last")
print("All tests passed!")
